dlls/nvcuda/nvcuda.c

```diff
@@ -3651,23 +3652,122 @@ CUresult WINAPI wine_cuDeviceGetUuid(CUuuid *uuid, CUdevice dev)
     else return CUDA_ERROR_INVALID_VALUE;
 }
 
```

Comment:
Would it be better - in case of frequent spam of this LUID function - to just obtain the LUID in DLL__PROCESS_ATTACH and set something like wine_luid[]; as a global parameter, and just use that in the cuDeviceGetLuid function?

Ofc that would mean instead of returning an error if UUID or registry is not correct, it would fallback to some set value like it is now.. Might not be a huge issue, but i somehow am inclined to think opening/closing registry carry some overhead.. I know this call is not expected to be used frequently, but well.. sometimes the cuda runtimes do strange things VERY frequently.

Reply:
We would have to store some kind of dictionary/map of UUIDs to LUIDs which I wanted to avoid because standard C library doesn't exactly help with that and I don't know if Wine has some builtin data structure for it like it does for linked lists.

Reply:
This part of registry is managed by https://gitlab.winehq.org/wine/wine/-/blob/wine-10.0-rc1/dlls/win32u/sysparams.c, which (afaict) writes GPUs as enumerated by display driver (e.g. winex11 or winewayland) in `add_gpu` / `write_gpu_to_registry` with remaining GPUs that weren't detected properly by display driver (due to incomplete implementation or xwayland limitations) being added from Vulkan, and I believe this is done whenever Wineserver is restarted but feel free to prove me wrong because I'm not sure about that part.

It's possible that existing LUID is kept as-is if Wine decides there's no need to update anything. Try deleting the key under `PCI` altogether and see if Wine regenerates it on prefix/wineserver restart.

> Hmm.. Is getting the UUID done for "searching purposes" in the registry?

Yes.

> I dont know how this looks in the registry in case of multiple adapters tbh.. (And i suppose this is the list you are talking about). So, if the UUID match -> copy the LUID? Guess this is the case if you have multiple adapters, and that is why UUID needs to be used.

Correct, this ensures we won't return LUID of another, possibly non-NV GPU. I can show you how it looks in Wine on my Intel/NV laptop (so, 2 GPUs, even if only one is CUDA-capable) in about ~6 hours from now.

Reply:
Yeah, i missed the usage of UUID at the beginning, so it seemed "simple enough to just get LUID" but needing to search using the actual current UUID it just was not useful to split this.. 

Thanks 😄

Reply:
![image](https://github.com/user-attachments/assets/4fe0f7a0-0e91-4e32-b77b-d80fd77b261b)

Here's how LUID for my Nvidia GPU is stored. Vulkan UUID is in the `{233A9EF3-AFC4-4ABD-B564-C32F21F1535C}\0002` property above and this pattern is repeated for my Intel GPU under `VEN_8086&DEV_A788&SUBSYS_00000000&REV_00` key.

Reply:
Ok.. so a `4080 Max-Q / Mobile`

Probably not really too interesting, but i do wonder how it would look in the case of 2 of the same card that would really have the same ven, dev and subsys version.. In windows the "subfolder" of 00000000 or 00000001 depending on (i guess) which adapter you have, have its own intricate unique number it seems - `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Enum\PCI\VEN_10DE&DEV_2705&SUBSYS_513E1462&REV_A1\4&23b96b75&0&0008`

Perhaps in Wine it would simply be `HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Enum\PCI\VEN_10DE&DEV_2705&SUBSYS_513E1462&REV_A1\00000000`
and
`HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Enum\PCI\VEN_10DE&DEV_2705&SUBSYS_513E1462&REV_A1\00000001`

and so on? 😄

Reply:
I don't know but it sounds pretty likely, Maybe it could also be observed when you have multiple Vulkan physical devices for a single GPU, for example RADV and AMDVLK supporting the same AMD GPU.